src/ORM/EagerLoadedDataList.php

```diff
@@ -0,0 +1,893 @@
+ */
+class EagerLoadedDataList extends ViewableData implements Relation, SS_List, Filterable, Sortable, Limitable
+{
+    /**
+     * List responsible for instantiating the actual DataObject objects from eager-loaded data
+     */
+    private DataList $dataList;
+
+    /**
+     * Underlying DataObject class for this list
+     */
+    private string $dataClass;
+
+    /**
+     * The ID(s) of the record that owns this list if the list represents a relation
+     * Used for aggregations
+     */
+    private int|array|null $foreignID;
+
+    /**
+     * ID-indexed array that holds the data from SQL queries for the list
+     * @var array<int,array>
+     */
+    private array $rows = [];
+
+    /**
+     * Nested eager-loaded data which applies to relations on records contained in this list
+     * @var array<int,self|DataObject>
+     */
+    private array $eagerLoadedData = [];
+
+    private array $extraFields = [];
+
+    private array $limitOffset = [null, 0];
+
+    private string|array $sort = [];
+
+    public function __construct(string $dataListClass, string $dataClass, int|array|null $foreignID = null, array $manyManyComponent = [])
+    {
+        if (!is_a($dataListClass, DataList::class, true)) {
+            throw new LogicException('$dataListClass must be an instanceof DataList');
+        }
+
+        // relation lists require a valid foreignID or set of IDs
+        if (is_a($dataListClass, RelationList::class, true) && !$this->isValidForeignID($foreignID)) {
+            throw new InvalidArgumentException('$foreignID must be a valid ID for eager loaded relation lists');
+        }
+
+        $this->dataClass = $dataClass;
+        $this->foreignID = $foreignID;
+
+        // many_many relation lists have extra constructor args that don't apply for has_many or non-relations
+        if (is_a($dataListClass, ManyManyThroughList::class, true)) {
+            $this->dataList = ManyManyThroughList::create(
+                $dataClass,
+                // If someone instantiates one of these and passes DataObjectSchema::manyManyComponent() directly
+                // the class will be in here as 'join'
+                $manyManyComponent['joinClass'] ?? $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields'],
+                $dataClass,
+                $manyManyComponent['parentClass']
+            );
+        } elseif (is_a($dataListClass, ManyManyList::class, true)) {
+            $this->dataList = ManyManyList::create(
+                $dataClass,
+                $manyManyComponent['join'],
+                $manyManyComponent['childField'],
+                $manyManyComponent['parentField'],
+                $manyManyComponent['extraFields']
+            );
+        } else {
+            $this->dataList = $dataListClass::create($dataClass, '');
```

Comment:
What's a scenario where a DataList that's not a RelationList is passed in for $dataListClass?

Reply:
It's used in testing, and could be used by dark wizards to similarly input data from who-knows-where to later hydrate some `DataObject` records.
The $dataListClass` is used to hydrate records from data, so if the data passed in isn't for a relation, then the `$dataListClass` should just be a `DataList`.